File: 1073-number-of-enclaves/1073-number-of-enclaves.cpp
class Solution {
public:
    int dfs(int i, int j ,int row, int col,vector<vector<int>>& grid,vector<vector<int>>& vis)
    {
        if(i<0 || j<0 || i==row||j==col || grid[i][j]==0 || vis[i][j]==1)
        {
            return 0;
        }
        vis[i][j]=1;
        int cnt =1 ;
        cnt += dfs(i+1,j,row,col,grid,vis)+dfs(i-1,j,row,col,grid,vis)+dfs(i,j+1,row,col,grid,vis)+dfs(i,j-1,row,col,grid,vis);
        return cnt;
    }
    int numEnclaves(vector<vector<int>>& grid) {
        int row = grid.size();
        int col = grid[0].size();
        int land =0;
        int bland =0;
        vector<vector<int>> vis(row,vector<int>(col,0));
        for(int i=0;i<row;i++)
        {
            for(int j =0;j<col;j++)
            {
                if(grid[i][j])
                    land++;   
                if(!vis[i][j] && grid[i][j]==1 && (i==0 || j==0 || i==row-1 || j==col-1))
                {
                   bland+= dfs(i,j,row,col,grid,vis);
                }
            }
        }
        return land-bland;
    }
};

File: 1073-number-of-enclaves/1073-number-of-enclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        row,col = len(grid) , len(grid[0])
        vis = set()
        land, bland = 0,0
        
        def dfs(i,j):
            if i < 0 or j<0 or i==row or j==col or (i,j) in vis or grid[i][j] == 0:
                return 0
            cnt = 1    
            vis.add((i,j))
            cnt += dfs(i+1,j)+dfs(i-1,j)+dfs(i,j+1)+dfs(i,j-1)
            return cnt

        for i in range(row):
            for j in range(col):
                land += grid[i][j]
                if grid[i][j]==1 and (i,j) not in vis and (i in [0,row-1] or j in [0,col-1]):
                    bland += dfs(i,j)


        return land - bland 
        